### PR TITLE
Fix crash in recursive instrumentation.

### PR DESCRIPTION
**Pull request by danielfenner**
The previous "fix" restored the return address in the entry payload. But
the return address was altered *after* the call to the entry payload. So
this never worked.

Removed the overwriting of the return address from the trampoline and
shifted it into the entry payload. This way we can decide if it is
appropriate to do the instrumentation or not depending on wether we are
re-entering the payload.

Bug: http://b/207361417
Test: Ran locally. Multiple times.

### CODE REVIEW

**Comment by beckerhe** [comment]
Awesome that you found the problem. Is it worthwhile/possible to add a test to TrampolineTest that's instrumenting clock_gettime?

File: src/UserSpaceInstrumentation/Trampoline.h [hunk 1]
```diff
@@ -116,18 +116,19 @@ struct RelocatedInstruction {
...
-// if the function shorter). `capstone_handle` is a handle to the capstone disassembler library
-// returned by cs_open. The function returns an error if it was not possible to instrument the
-// function. For details on that see the comments at AppendRelocatedPrologueCode. If the function is
-// successful it will insert an address pair into `relocation_map` for each instruction it relocated
-// from the beginning of the function into the trampoline (needed for moving instruction pointers
-// away from the overwritten bytes at the beginning of the function, compare
-// MoveInstructionPointersOutOfOverwrittenCode below). The return value is the address of the first
-// instruction not relocated into the trampoline (i.e. the address the trampoline jump back to).
+// function's return address, a function id, the address on the stack where the return address is
+// stored and and the address of the return trampoline as parameters. The function id is written
```
Comment (dpallotti) [inline src/UserSpaceInstrumentation/Trampoline.h:120]:
```suggestion
// stored, and the address of the return trampoline as parameters. The function id is written
```

**Review by dpallotti — APPROVED** [review]

File: src/UserSpaceInstrumentation/Trampoline.h (hunk 1 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/Trampoline.h:120]: done

**Comment by danielfenner** [comment]
> Awesome that you found the problem. Is it worthwhile/possible to add a test to TrampolineTest that's instrumenting clock_gettime?

I had the test more or less finished but didn't get it to "work" (i.e. crash when this fix is not applied) as symbols in the libc are different on my development system and the target system. I'll add something to the integration test.

**Review by beckerhe — APPROVED** [review]

**Pull request merged**